src/cmd/services/m3coordinator/downsample/downsampler.go

```diff
@@ -0,0 +1,740 @@
+	"github.com/m3db/m3metrics/filters"
+	"github.com/m3db/m3metrics/matcher"
+	"github.com/m3db/m3metrics/matcher/cache"
+	"github.com/m3db/m3metrics/metric/aggregated"
+	"github.com/m3db/m3metrics/metric/id"
+	"github.com/m3db/m3metrics/metric/unaggregated"
+	"github.com/m3db/m3metrics/policy"
+	"github.com/m3db/m3metrics/rules"
+	"github.com/m3db/m3x/clock"
+	xerrors "github.com/m3db/m3x/errors"
+	"github.com/m3db/m3x/ident"
+	"github.com/m3db/m3x/instrument"
+	"github.com/m3db/m3x/pool"
+	xsync "github.com/m3db/m3x/sync"
+
+	"github.com/coreos/etcd/integration"
+	"github.com/uber-go/tally"
+)
+
+const (
+	initAllocTagsSliceCapacity     = 32
+	shardSetID                     = uint32(0)
+	instanceID                     = "downsampler_local"
+	placementKVKey                 = "/placement"
+	aggregationSuffixTag           = "aggregation"
+	defaultStorageFlushConcurrency = 20000
+)
+
+var (
+	numShards = runtime.NumCPU()
+
+	errNoStorage               = errors.New("dynamic downsampling enabled with storage set")
+	errNoRulesStore            = errors.New("dynamic downsampling enabled with rules store not set")
+	errNoClockOptions          = errors.New("dynamic downsampling enabled with clock options not set")
+	errNoInstrumentOptions     = errors.New("dynamic downsampling enabled with instrument options not set")
+	errNoTagEncoderOptions     = errors.New("dynamic downsampling enabled with tag encoder options not set")
+	errNoTagDecoderOptions     = errors.New("dynamic downsampling enabled with tag decoder options not set")
+	errNoTagEncoderPoolOptions = errors.New("dynamic downsampling enabled with tag encoder pool options not set")
+	errNoTagDecoderPoolOptions = errors.New("dynamic downsampling enabled with tag decoder pool options not set")
+)
+
+// Downsampler is a downsampler.
+type Downsampler interface {
+	MetricsAppender() MetricsAppender
+}
+
+// MetricsAppender is a metrics appender that can
+// build a samples appender, only valid to use
+// with a single caller at a time.
+type MetricsAppender interface {
+	AddTag(name, value string)
+	SamplesAppender() (SamplesAppender, error)
+	Reset()
+	Finalize()
+}
+
+// SamplesAppender is a downsampling samples appender,
+// that can only be called by a single caller at a time.
+type SamplesAppender interface {
+	AppendCounterSample(value int64) error
+	AppendGaugeSample(value float64) error
+}
+
+// DownsamplerOptions is a set of required downsampler options.
+type DownsamplerOptions struct {
+	Storage                 storage.Storage
+	StorageFlushConcurrency int
+	RulesKVStore            kv.Store
+	ClockOptions            clock.Options
+	InstrumentOptions       instrument.Options
+	TagEncoderOptions       serialize.TagEncoderOptions
+	TagDecoderOptions       serialize.TagDecoderOptions
+	TagEncoderPoolOptions   pool.ObjectPoolOptions
+	TagDecoderPoolOptions   pool.ObjectPoolOptions
+}
+
+// Validate will validate the dynamic downsampling options.
+func (o DownsamplerOptions) Validate() error {
```

Comment:
nit: Should Validate() have the responsibility of checking that `o.StorageFlushConcurrency > 0`, and setting it to default if not?

Reply:
Sure thing, I'll remove visibility.

Reply:
Also: validate steps should never mutate something, hence I'm pretty against mutating during validate.